gcp_monitor.py

```diff
@@ -395,9 +429,44 @@ def get_time_series(gcp_variables, metric_descriptor, value):
     return series
 
 
-def report(gcp_variables, metrics_client):
+def report(gcp_variables, metrics_client, nvml_ok: bool = False):
 
+    num_gpus = pynvml.nvmlDeviceGetCount() if nvml_ok else 0
     time_delta = time() - gcp_variables["last_time"]
+    gpus = [pynvml.nvmlDeviceGetHandleByIndex(i) for i in range(num_gpus)]
+    # https://pypi.org/project/nvidia-ml-py is a thin wrapper over the C NVML library
+    # See https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html
+    # for type info and field names
+    gpu_utilization_rates = [
+        pynvml.nvmlDeviceGetUtilizationRates(gpus[i]) for i in range(num_gpus)
+    ]
+    gpu_mem_info = [pynvml.nvmlDeviceGetMemoryInfo(gpus[i]) for i in range(num_gpus)]
+    gpu_metrics = [
+        *[
+            get_time_series(
+                gcp_variables,
+                gcp_variables[f"GPU{i}_UTILIZATION_METRIC"],
+                {"int64_value": (gpu_utilization_rates[i].gpu)},
+            )
+            for i in range(num_gpus)
+        ],
+        *[
+            get_time_series(
+                gcp_variables,
+                gcp_variables[f"GPU{i}_MEM_UTILIZATION_METRIC"],
+                {"int64_value": (gpu_utilization_rates[i].memory)},
+            )
+            for i in range(num_gpus)
+        ],
+        *[
+            get_time_series(
+                gcp_variables,
+                gcp_variables[f"GPU{i}_MEM_USED_METRIC"],
+                {"int64_value": (gpu_mem_info[i].memory)},
```

Comment:
```suggestion
                {"int64_value": (gpu_mem_info[i].used)},
```
Correct me if I'm wrong, but I'm pretty sure the documentation you linked says you should be accessing the fields within the returned class. memory doesn't seem to be one of those fields
https://docs.nvidia.com/deploy/nvml-api/structnvmlMemory__t.html#structnvmlMemory__t

Reply:
```suggestion
                {"int64_value": 100 * (gpu_mem_info[i].used)/(gpu_mem_info[i].total)},
```
The function actually doesn't even seem to return a percentage, so we need to calulate it. Alternatively use integer division if you want to keep this metric as an int.

Reply:
> memory doesn't seem to be one of those fields

concerning that my verification didn't trip on this. Will reverify and fix.

Reply:
pffff I guess I didn't verify after making some changes because now I get an error
```
Traceback (most recent call last):
  File "/opt/./monitor.py", line 66, in <module>
    main()
  File "/opt/./monitor.py", line 54, in main
    gcp_instance = gcp_monitor.report(gcp_instance, metrics_client, nvml_ok)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/gcp_monitor.py", line 461, in report
    {"int64_value": (gpu_mem_info[i].memory)},
                     ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pynvml.py", line 1076, in __getattribute__
    res = super(_PrintableStructure, self).__getattribute__(name)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'c_nvmlMemory_t' object has no attribute 'memory'
```